v2/mongodb-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/mongodb/templates/JavascriptDocumentTransformer.java

```diff
@@ -0,0 +1,267 @@
+/*
+ * Copyright (C) 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.teleport.v2.mongodb.templates;
+
+import static org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions.checkArgument;
+
+import com.google.auto.value.AutoValue;
+import java.io.IOException;
+import java.io.Reader;
+import java.io.UncheckedIOException;
+import java.nio.channels.Channels;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import javax.script.Invocable;
+import javax.script.ScriptEngine;
+import javax.script.ScriptEngineFactory;
+import javax.script.ScriptEngineManager;
+import javax.script.ScriptException;
+import jdk.nashorn.api.scripting.ScriptObjectMirror;
+import org.apache.beam.sdk.io.FileSystems;
+import org.apache.beam.sdk.io.fs.MatchResult;
+import org.apache.beam.sdk.io.fs.MatchResult.Metadata;
+import org.apache.beam.sdk.io.fs.MatchResult.Status;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Strings;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.io.CharStreams;
+import org.bson.Document;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** A Document UDF Transform Function. Note that this class's implementation is not threadsafe */
+@AutoValue
+public abstract class JavascriptDocumentTransformer {
```

Comment:
Could this not extend JavascriptTextTransformer? It could also be put into the v2/common/.../transforms folder alongside JavascriptTextTransformer.java to be reused if there is ever another MongoDB template that uses UDF's.

Note: You will also have to remove private from the getJavascriptRuntime() method in JavascriptTextTransformer.java

```
@AutoValue
public class JavascriptDocumentTransformer extends JavascriptTextTransformer {
  public interface JavascriptTextTransformerOptions extends PipelineOptions {}

  @AutoValue
  public abstract static class JavascriptRuntime extends JavascriptTextTransformer.JavascriptRuntime {
    @Nullable
    public Document invoke(Document data)
        throws ScriptException, IOException, NoSuchMethodException {
      Invocable invocable = getInvocable();
      if (invocable == null) {
        throw new RuntimeException("No udf was loaded");
      }

      Object result = getInvocable().invokeFunction(functionName(), data);
      if (result == null || ScriptObjectMirror.isUndefined(result)) {
        return null;
      } else if (result instanceof Document) {
        return (Document) result;
      } else {
        String className = result.getClass().getName();
        throw new RuntimeException(
            "UDF Function did not return a String. Instead got: " + className);
      }
    }
  }

  @AutoValue
  public abstract static class TransformDocumentViaJavascript
      extends PTransform<PCollection<Document>, PCollection<Document>> {
    public abstract @Nullable String fileSystemPath();

    public abstract @Nullable String functionName();

    /** Builder for {@link TransformDocumentViaJavascript}. */
    @AutoValue.Builder
    public abstract static class Builder {
      public abstract Builder setFileSystemPath(@Nullable String fileSystemPath);

      public abstract Builder setFunctionName(@Nullable String functionName);

      public abstract TransformDocumentViaJavascript build();
    }

    public static Builder newBuilder() {
      return new AutoValue_JavascriptDocumentTransformer_TransformDocumentViaJavascript.Builder();
    }

    @Override
    public PCollection<Document> expand(PCollection<Document> doc) {
      return doc.apply(
          ParDo.of(
              new DoFn<Document, Document>() {
                private JavascriptRuntime javascriptRuntime;

                @Setup
                public void setup() {
                  if (fileSystemPath() != null && functionName() != null) {
                    javascriptRuntime = (JavascriptRuntime) getJavascriptRuntime(fileSystemPath(), functionName());
                  }
                }

                @ProcessElement
                public void processElement(ProcessContext c)
                    throws IOException, NoSuchMethodException, ScriptException {
                  Document element = c.element();
                  if (javascriptRuntime != null) {
                    element = javascriptRuntime.invoke(element);
                    c.output(element);
                  } else {
                    c.output(element);
                  }
                }
              }));
    }
  }
}
```

Reply:
Moved the file to common/transforms

I will update the rest suggestion in next PR. We are waiting to release some of the updates asap and refactoring requires some time and testing.

Reply:
Sounds good.